staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -906,3 +908,107 @@ func TestDispatchEventWillNotBeBlockedByTimedOutWatcher(t *testing.T) {
 		t.Errorf("watcher is blocked by slower one (count: %d)", eventsCount)
 	}
 }
+
+func verifyEvents(t *testing.T, w watch.Interface, events []watch.Event) {
+	_, _, line, _ := goruntime.Caller(1)
+	for _, expectedEvent := range events {
+		select {
+		case event := <-w.ResultChan():
+			if e, a := expectedEvent.Type, event.Type; e != a {
+				t.Logf("(called from line %d)", line)
+				t.Errorf("Expected: %s, got: %s", e, a)
+			}
+			object := event.Object
+			if co, ok := object.(runtime.CacheableObject); ok {
+				object = co.GetObject()
+			}
+			if e, a := expectedEvent.Object, object; !apiequality.Semantic.DeepEqual(e, a) {
+				t.Logf("(called from line %d)", line)
+				t.Errorf("Expected: %#v, got: %#v", e, a)
+			}
+		case <-time.After(wait.ForeverTestTimeout):
+			t.Logf("(called from line %d)", line)
+			t.Errorf("Timed out waiting for an event")
+		}
+	}
+}
+
+func TestCachingDeleteEvents(t *testing.T) {
+	backingStorage := &dummyStorage{}
+	cacher, _, err := newTestCacher(backingStorage, 1000)
+	if err != nil {
+		t.Fatalf("Couldn't create cacher: %v", err)
+	}
+	defer cacher.Stop()
+
+	// Wait until cacher is initialized.
+	cacher.ready.wait()
+
+	fooPredicate := storage.SelectionPredicate{
+		Label: labels.SelectorFromSet(map[string]string{"foo": "true"}),
+		Field: fields.Everything(),
+	}
+	barPredicate := storage.SelectionPredicate{
+		Label: labels.SelectorFromSet(map[string]string{"bar": "true"}),
+		Field: fields.Everything(),
+	}
+
+	createWatch := func(pred storage.SelectionPredicate) watch.Interface {
+		w, err := cacher.Watch(context.TODO(), "pods/ns", "999", pred)
+		if err != nil {
+			t.Fatalf("Failed to create watch: %v", err)
+		}
+		return w
+	}
+
+	w1 := createWatch(storage.Everything)
+	defer w1.Stop()
+	w2 := createWatch(fooPredicate)
+	defer w2.Stop()
+	w3 := createWatch(barPredicate)
+	defer w3.Stop()
+
+	makePod := func(labels map[string]string, rv string) *examplev1.Pod {
+		return &examplev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "pod",
+				Namespace:       "ns",
+				Labels:          labels,
+				ResourceVersion: rv,
+			},
+		}
+	}
+	pod1 := makePod(map[string]string{"foo": "true", "bar": "true"}, "1001")
+	cacher.watchCache.Add(pod1)
+	pod2 := makePod(map[string]string{"foo": "true"}, "1002")
+	cacher.watchCache.Update(pod2)
+	pod3 := makePod(map[string]string{}, "1003")
+	cacher.watchCache.Update(pod3)
+	pod4 := makePod(map[string]string{}, "1004")
+	cacher.watchCache.Delete(pod4)
+
+	w1Events := []watch.Event{
+		{Type: watch.Added, Object: pod1},
+		{Type: watch.Modified, Object: pod2},
+		{Type: watch.Modified, Object: pod3},
+		{Type: watch.Deleted, Object: pod4},
+	}
+	verifyEvents(t, w1, w1Events)
+
+	pod2DeletedAt3 := pod2.DeepCopyObject().(*examplev1.Pod)
+	pod2DeletedAt3.ResourceVersion = "1003"
```

Comment:
initialize all pods (including the podNDeletedAt pods) before we pass them to any watch cache methods to ensure no modifications are made by the watch cache and bleed into comparisons

similarly, initialize all expected watch event slices with copies of the pods before passing the pods into any watch cache methods

```
pod1 := ...
pod2 := ...
pod3 := ...
pod4 := ...
pod1DeletedAt2 := ...
pod2DeletedAt3 := ...
everythingWatcherEvents := []watch.Event{
  {Type: watch.Added, Object: pod1.DeepCopy()},
  ...
}
fooWatcherEvents := []watch.Event{
  ...
}
barWatcherEvents := []watch.Event{
  ...
}
```

Reply:
done